storage/range.go

```diff
@@ -819,23 +852,37 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
 			if log.V(1) {
 				log.Infoc(ctx, "found response cache entry for %+v", args.Header().CmdID)
 			}
-			return reply.Header().GoError()
+			return nil, err
 		} else if ok && err != nil {
-			newErr := util.Errorf("unable to read result for %+v from the response cache: %s", args, err)
-			reply.Header().SetGoError(newErr)
-			return newErr
+			return nil, newReplicaCorruptionError(
+				util.Errorf("could not read from response cache"), err)
 		}
 	}
 
-	// Verify the leader lease is held; Note that we don't require the
-	// leader lease when trying to grant the leader lease!
-	if _, ok := args.(*proto.InternalLeaderLeaseRequest); !ok {
-		if !r.verifyLeaderLease(originNodeID, header.Timestamp) {
-			err := r.newNotLeaderError()
-			reply.Header().SetGoError(err)
-			return err
+	committed := false
+
+	defer func() {
+		if !committed {
+			// We didn't commit the batch, but advance the last applied index nonetheless.
```

Comment:
This should come before the read from the response cache. We should advance the applied index even if we "applied" the command by reading a cached response. In fact, we also need to advance the applied index if we're returning a notLeaderError (since the retried command will be at a new log index)